File: photonflow-cpp/src/blocks/optical/laser.cpp
/**
 * @file laser.cpp
 * @brief Laser source block implementation.
 */

#include "photonflow/blocks/optical/laser.hpp"
#include "photonflow/blocks/block_registry.hpp"

#include <cmath>
#include <complex>

namespace photonflow {

Laser::Laser(const std::string &id, const json &params, const json &nonideal)
    : BaseBlock(id, params, nonideal) {}

std::unordered_map<std::string, Signal>
Laser::process(const std::unordered_map<std::string, Signal> & /*inputs*/,
               SimContext &ctx) {
  // Extract parameters
  double power_dbm = get_param("power_dbm", 0.0);
  double center_freq = get_param("center_freq_hz", 193.1e12);
  double phase0 = get_param("phase0", 0.0);

  // Convert power to linear scale
  double power_w = 1e-3 * std::pow(10.0, power_dbm / 10.0);
  double amp = std::sqrt(power_w);

  // Check if non-ideal effects are enabled
  bool nonideal_enable = get_nonideal("enable", false);

  if (nonideal_enable) {
    // Power error
    double power_error_db = get_nonideal("power_error_db", 0.0);
    if (power_error_db != 0.0) {
      amp *= std::pow(10.0, power_error_db / 20.0);
    }

    // Frequency offset
    double freq_offset = get_nonideal("freq_offset_hz", 0.0);
    center_freq += freq_offset;
  }

  const int n = ctx.n_samples();
  Eigen::VectorXcd data(n);

  if (nonideal_enable) {
    double linewidth = get_nonideal("linewidth_hz", 0.0);
    double phase_noise_rms = get_nonideal("phase_noise_rms", 0.0);
    double rin_db = get_nonideal("rin_db_per_hz", -150.0);

    // Generate phase array
    Eigen::VectorXd phase = Eigen::VectorXd::Constant(n, phase0);

    // Add random phase noise
    if (phase_noise_rms > 0.0) {
      Eigen::VectorXd noise = ctx.randn(n);
      phase += phase_noise_rms * noise;
    }

    // Add linewidth-induced phase noise (Wiener process)
    if (linewidth > 0.0) {
      double sigma = std::sqrt(2.0 * M_PI * linewidth / ctx.fs());
      Eigen::VectorXd dphi = sigma * ctx.randn(n);
      // Cumulative sum
      for (int i = 1; i < n; ++i) {
        phase[i] = phase[i] + phase[i - 1] + dphi[i];
      }
      phase[0] += dphi[0];
    }

    // Handle RIN (Relative Intensity Noise)
    if (rin_db > -200.0) {
      double rin_linear = std::pow(10.0, rin_db / 10.0);
      double sigma_i = std::sqrt(rin_linear * ctx.fs() / 2.0);
      Eigen::VectorXd intensity_noise = ctx.randn(n);

      for (int i = 0; i < n; ++i) {
        double intensity = power_w * (1.0 + sigma_i * intensity_noise[i]);
        intensity = std::max(intensity, 0.0);
        double amp_i = std::sqrt(intensity);
        data[i] = amp_i * std::exp(std::complex<double>(0.0, phase[i]));
      }
    } else {
      // No RIN
      for (int i = 0; i < n; ++i) {
        data[i] = amp * std::exp(std::complex<double>(0.0, phase[i]));
      }
    }
  } else {
    // Ideal case: pure CW with constant amplitude and phase
    std::complex<double> field =
        amp * std::exp(std::complex<double>(0.0, phase0));
    data.setConstant(field);
  }

  Signal output(data, ctx.fs(), ctx.t0());
  output.center_freq = center_freq;
  output.pol_mode = "scalar";

  return {{"opt_out", std::move(output)}};
}

std::optional<std::string> Laser::port_type(const std::string &port) const {
  if (port == "opt_out")
    return "optical";
  return std::nullopt;
}

json Laser::describe() const {
  return {
      {"ports", {{"opt_out", "optical"}}},
      {"spec",
       {{"params",
         {{"power_dbm", {{"type", "float"}, {"default", 0.0}, {"unit", "dBm"}}},
          {"center_freq_hz",
           {{"type", "float"}, {"default", 193.1e12}, {"unit", "Hz"}}},
          {"phase0", {{"type", "float"}, {"default", 0.0}, {"unit", "rad"}}}}},
        {"nonideal",
         {{"enable", {{"type", "bool"}, {"default", false}}},
          {"linewidth_hz",
           {{"type", "float"}, {"default", 0.0}, {"unit", "Hz"}}},
          {"rin_db_per_hz",
           {{"type", "float"}, {"default", -150.0}, {"unit", "dB/Hz"}}},
          {"power_error_db",
           {{"type", "float"}, {"default", 0.0}, {"unit", "dB"}}},
          {"freq_offset_hz",
           {{"type", "float"}, {"default", 0.0}, {"unit", "Hz"}}},
          {"phase_noise_rms",
           {{"type", "float"}, {"default", 0.0}, {"unit", "rad"}}}}}}}};
}

// Auto-register
REGISTER_BLOCK(Laser, "Laser");

} // namespace photonflow

File: backend/src/photonflow/blocks/optical/laser.py
"""Laser source block."""

from __future__ import annotations

import math
from typing import Dict

import torch

from photonflow.blocks.base import BaseBlock, register_block
from photonflow.core.signal import Signal
from photonflow.core.sim import SimContext


@register_block("Laser")
class Laser(BaseBlock):
    """Continuous-wave (CW) laser source. Generates an optical carrier signal with configurable power, frequency, and phase. Supports linewidth, RIN, and other non-ideal effects."""

    PORTS = {"opt_out": "optical"}
    SPEC = {
        "params": {
            "power_dbm": {"type": "float", "default": 0.0, "unit": "dBm"},
            "center_freq_hz": {"type": "float", "default": 193.1e12, "unit": "Hz"},
            "phase0": {"type": "float", "default": 0.0, "unit": "rad"},
        },
        "nonideal": {
            "enable": {"type": "bool", "default": False},
            "linewidth_hz": {"type": "float", "default": 0.0, "unit": "Hz"},
            "rin_db_per_hz": {"type": "float", "default": -150.0, "unit": "dB/Hz"},
            "power_error_db": {"type": "float", "default": 0.0, "unit": "dB"},
            "freq_offset_hz": {"type": "float", "default": 0.0, "unit": "Hz"},
            "phase_noise_rms": {"type": "float", "default": 0.0, "unit": "rad"},
        },
    }

    def process(self, inputs: Dict[str, Signal], ctx: SimContext) -> Dict[str, Signal]:
        power_dbm = float(self.params.get("power_dbm", 0.0))
        center_freq = float(self.params.get("center_freq_hz", 193.1e12))
        phase0 = float(self.params.get("phase0", 0.0))

        power_w = 1e-3 * (10 ** (power_dbm / 10.0))
        amp = math.sqrt(power_w)

        t = ctx.time()
        dtype = torch.complex64
        phase = phase0 + torch.zeros_like(t)

        nonideal = self.nonideal if self.nonideal.get("enable", False) else {}
        power_error_db = float(nonideal.get("power_error_db", 0.0))
        if power_error_db != 0.0:
            amp = amp * (10 ** (power_error_db / 20.0))
        freq_offset = float(nonideal.get("freq_offset_hz", 0.0))
        if freq_offset != 0.0:
            center_freq = center_freq + freq_offset
        phase_noise_rms = float(nonideal.get("phase_noise_rms", 0.0))
        if phase_noise_rms > 0.0:
            phase = phase + phase_noise_rms * torch.randn_like(t)
        linewidth = float(nonideal.get("linewidth_hz", 0.0))
        if linewidth > 0.0:
            sigma = math.sqrt(2.0 * math.pi * linewidth / ctx.fs)
            dphi = sigma * torch.randn_like(t)
            phase = phase + torch.cumsum(dphi, dim=0)

        rin_db = nonideal.get("rin_db_per_hz")
        if rin_db is not None:
            rin_linear = 10 ** (float(rin_db) / 10.0)
            sigma_i = math.sqrt(rin_linear * ctx.fs / 2.0)
            intensity = power_w * (1.0 + sigma_i * torch.randn_like(t))
            intensity = torch.clamp(intensity, min=0.0)
            amp = torch.sqrt(intensity)

        field = amp * torch.exp(1j * phase.to(dtype=dtype))
        signal = Signal(data=field, fs=ctx.fs, t0=ctx.t0, center_freq=center_freq)
        return {"opt_out": signal}
